non-tradable-token (#4671)
ethereum/EIPs: EIPS/eip-4671.md

---
eip: 4671
title: Badge Standard
description: A standard interface for badges.
author: Omar Aflak (@omaraflak), Pol-Malo Le Bris, Marvin Martin (@MarvinMartin24)
discussions-to: https://ethereum-magicians.org/t/eip-4671-non-tradable-token/7976
status: Draft
type: Standards Track
category: ERC
created: 2022-01-13
requires: 165
---

> # Abstract

Reviewer: ```suggestion
## Abstract
```
The title hierarchy appears to be off by one. Please fix this throughout the EIP.

Author: Thank you for taking the time to look!

    /// @param badge Address of the IERC4671Enumerable contract to remove
    function remove(address badge) external;

    /// @notice Get all the IERC4671Enumerable contracts for a given owner
    /// @param owner Address for which to retrieve the IERC4671Enumerable contracts
    function get(address owner) external view returns (address[] memory);
}
```

## Rationale

### On-chain vs Off-chain

A decision was made to keep the data off-chain (via `badgeURI()`) for two main reasons: 
* Badges represent personal possessions. Therefore, there might be cases where the data should be encrypted. The standard should not outline decisions about encryption because there are just so many ways this could be done, and every possibility is specific to the use-case.
* Badges must stay generic. There could have been a possibility to make a `MetadataStore` holding the data of badges in an elegant way, unfortunately we would have needed a support for generics in solidity (or struct inheritance), which is not available today.

## Reference Implementation

You can find implementations of the badge standard at the following links.

> * [https://github.com/OmarAflak/ERC4671/tree/master/contracts](https://github.com/OmarAflak/ERC4671/tree/master/contracts)

Reviewer 2: This should be inlined or add as an asset at `../assets/eip-4671/*`.

---
eip: 4671
title: Non-Tradable Tokens
description: A standard interface for non-tradable tokens.
author: Omar Aflak (@omaraflak), Pol-Malo Le Bris, Marvin Martin (@MarvinMartin24)
discussions-to: https://ethereum-magicians.org/t/eip-4671-non-tradable-token/7976
status: Draft
type: Standards Track
category: ERC
created: 2022-01-13
requires: 165
---

## Abstract

> NTTs represent inherently personal possessions (material or immaterial), such as university diploma, online training certificates, government issued documents (national id, driving licence, visa, wedding, etc.), badges, labels, and so on.

Author: Thanks for the typo!
Haha yes I guess ^^" Any ideas ? I was thinking "Non-Exchangeable Token" => NXT ?

> NTTs represent inherently personal possessions (material or immaterial), such as university diploma, online training certificates, government issued documents (national id, driving licence, visa, wedding, etc.), badges, labels, and so on.

Author: Or maybe `Personal Token` ?

> NTTs represent inherently personal possessions (material or immaterial), such as university diploma, online training certificates, government issued documents (national id, driving licence, visa, wedding, etc.), badges, labels, and so on.

Author: I've opted for `Badge Standard`

    /// @param ntt Address of the NTT contract to add
    function add(address ntt) external;

    /// @notice Remove a NTT contract from the caller's record
    /// @param ntt Address of the NTT contract to remove
    function remove(address ntt) external;

    /// @notice Get all the NTT contracts for a given owner
    /// @param owner Address for which to retrieve the NTT contracts
    function get(address owner) external view returns (address[] memory);
}
```
<!-- AUTO-GENERATED-CONTENT:END -->

## Rationale

### On-chain vs Off-chain

A decision was made to keep the data off-chain (via `tokenURI()`) for two main reasons: 
* Non-Tradable Tokens represent personal possessions. Therefore, there might be cases where the data should be encrypted. The standard should not outline decisions about encryption because there are just so many ways this could be done, and every possibility is specific to the use-case.
* Non-Tradable Tokens must stay generic. There could have been a possibility to make a `MetadataStore` holding the data of NTTs in an elegant way, unfortunately we would have needed a support for generics in solidity (or struct inheritance), which is not available today.

> ## Implementation

Reviewer: ```suggestion
## Reference Implementation
```

Author: Thank you for taking the time to review!

    /// @param owner Address for whom to mint the token
    function approveMint(address owner) external;

    /// @notice Cast a vote to invalidate a specific token
    /// @param tokenId Identifier of the token to invalidate
    function approveInvalidate(uint256 tokenId) external;
}
```

##### Pull

An interface allowing a token owner to pull his token to a another of his wallets (here `recipient`). The caller must provide a signature of the tuple `(tokenId, owner, recipient)` using the `owner` wallet.

```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

import "./IERC4671.sol";

interface IERC4671Pull is IERC4671 {
    // Event emitted when a IERC4671 token is transfered from owner to recipient
    event Pulled(address owner, address recipient);

    /// @notice Pull a token from the owner wallet to the caller's wallet
    /// @param tokenId Identifier of the token to transfer
    /// @param owner Address that owns tokenId
    /// @param signature Signed data (tokenId, owner, recipient) by the owner of the token
    function pull(uint256 tokenId, address owner, bytes memory signature) external;
}

> ```

Author: As suggested when we last talked, I have added an extension that allows self-transfer of the token.

It's a "pull" mechanism, where you have to provide a signed message by the wallet that owns the token in order to transfer it to the caller (which would imply that the caller has access to the current owner's wallet).

    /// @notice Remove a IERC4671Enumerable contract from the caller's record
    /// @param token Address of the IERC4671Enumerable contract to remove
    function remove(address token) external;

    /// @notice Get all the IERC4671Enumerable contracts for a given owner
    /// @param owner Address for which to retrieve the IERC4671Enumerable contracts
    function get(address owner) external view returns (address[] memory);
}
```

## Rationale

### On-chain vs Off-chain

A decision was made to keep the data off-chain (via `tokenURI()`) for two main reasons: 
* Non-tradable tokens represent personal possessions. Therefore, there might be cases where the data should be encrypted. The standard should not outline decisions about encryption because there are just so many ways this could be done, and every possibility is specific to the use-case.
* Non-tradable tokens must stay generic. There could have been a possibility to make a `MetadataStore` holding the data of tokens in an elegant way, unfortunately we would have needed a support for generics in solidity (or struct inheritance), which is not available today.

## Reference Implementation

> You can find an implementation of this standard in [../assets/eip-4671](https://github.com/ethereum/EIPs/tree/master/assets/eip-4671).

Reviewer 2: ```suggestion
You can find an implementation of this standard [here](../assets/eip-4671).
```
Feel free to change the title text, but the link should be relative.

---
eip: 4671
title: Non-Tradable Tokens Standard
description: A standard interface for non-tradable tokens, aka badges or souldbound NFTs.
author: Omar Aflak (@omaraflak), Pol-Malo Le Bris, Marvin Martin (@MarvinMartin24)
discussions-to: https://ethereum-magicians.org/t/eip-4671-non-tradable-token/7976

> status: Review

Reviewer: ```suggestion
status: Draft
```

---
eip: 4671
title: Non-Tradable Tokens
description: A standard interface for non-tradable tokens.

> author: Omar Aflak (@omaraflak)
> contributors: Pol-Malo Le Bris, Marvin Martin (@MarvinMartin24)

Reviewer 2: ```suggestion
author: Omar Aflak (@omaraflak), Pol-Malo Le Bris, Marvin Martin (@MarvinMartin24)
```
```suggestion
author: Omar Aflak (@omaraflak)
```
Either is acceptable.  `contributors` isn't a valid header field though.

> discussions-to: https://github.com/ethereum/EIPs/issues/4672

Reviewer 2: This should be a forum thread at https://ethereum-magicians.org/

Author: Thank you for the comments!

---
eip: 4671
title: Non-Tradable Tokens
description: A standard interface for non-tradable tokens.
author: Omar Aflak (@omaraflak), Pol-Malo Le Bris, Marvin Martin (@MarvinMartin24)

> discussions-to: https://ethereum-magicians.org/t/eip-4671-non-tradable-token/7976?u=omaraflak

Reviewer 3: ```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-4671-non-tradable-token/7976
```

status: Draft
type: Standards Track
category: ERC
created: 2022-01-13
requires: 165
---

> # Non-Tradable Token Standard

Reviewer 3: ```suggestion
```
The EIP should only have sections mentioned in [EIP-1](./eip-1.md), so please remove this

> <!-- AUTO-GENERATED-CONTENT:START (TOC) -->
> - [Simple Summary](#simple-summary)
> - [Abstract](#abstract)
> - [Motivation](#motivation)
> - [Specification](#specification)
>   - [Extensions](#extensions)
>     - [Metadata](#metadata)
>     - [Delegation](#delegation)
> - [Rationale](#rationale)
>   - [On-chain vs Off-chain](#on-chain-vs-off-chain)
> - [Implementation](#implementation)
>   - [NTT](#ntt)
>   - [NTTDelegate](#nttdelegate)
> - [NTT for EIP ?](#ntt-for-eip-)
> - [Copyright](#copyright)
> <!-- AUTO-GENERATED-CONTENT:END -->
> 

Reviewer 3: ```suggestion
```
Even though the toc is useful. since it isn't in EIP-1, it should be removed

> ## Simple Summary
> 
> A standard interface for <u>**non-tradable tokens**</u>, aka <u>**NTT**</u>s.
> 

Reviewer 3: ```suggestion
```
The `Simple Summary` section has been removed (in favor of the `description` section in the header)

By providing a common interface for this type of tokens, we allow more applications to be developed and we position blockchain technology as a standard gateway for verification of personal possessions.

## Specification

A single NTT contract, is seen as representing one type of badge by one authority. For instance, one NTT contract for PSN achievements, another for Ethereum EIP creators, and so on...

* An address might possess multiple tokens, which are indexed.
* An authority who delivers a certificate should be in position to invalidate it. Think of driving licences or weddings. However, it cannot delete your token.
* The issuer of a token might be someone else than the contract creator.

<!-- AUTO-GENERATED-CONTENT:START (CODE:syntax=solidity&src=./contracts/INTT.sol) -->
<!-- The below code snippet is automatically added from ./contracts/INTT.sol -->
```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

> import "@openzeppelin/contracts/utils/introspection/IERC165.sol";

Reviewer 3: ```suggestion
import "./IERC165.sol";
```
I think it best not to reference specific implementations, and have seen other standards use a pseudocode implementation like the one I've suggested to represent inheritance (for example, [ERC-721](./eip-721.md)).

    /// @param owners Addresses for whom the tokens are minted
    function mintBatch(address[] memory owners) external;
}
```
<!-- AUTO-GENERATED-CONTENT:END -->

## Rationale

### On-chain vs Off-chain

A decision was made to keep the data off-chain (via `tokenURI()`) for two main reasons: 
* Non-Tradable Tokens represent personal possessions. Therefore, there might be cases where the data should be encrypted. The standard should not outline decisions about encryption because there are just so many ways this could be done, and every possibility is specific to the use-case.
* Non-Tradable Tokens must stay generic. There could have been a possibility to make a `MetadataStore` holding the data of NTTs in an elegant way, unfortunately we would have needed a support for generics in solidity (or struct inheritance), which is not available today.

## Implementation

### NTT

<!-- AUTO-GENERATED-CONTENT:START (CODE:syntax=solidity&src=./contracts/NTT.sol) -->
<!-- The below code snippet is automatically added from ./contracts/NTT.sol -->
```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

import "@openzeppelin/contracts/utils/Strings.sol";

> import "@openzeppelin/contracts/utils/introspection/ERC165.sol";

Author: I can't keep it even if it is in the `implementation` section and not `specification` ?

> import "@openzeppelin/contracts/utils/introspection/ERC165.sol";

Author: Shall I upload the implementation on a separate repo and simply reference it here ?

> import "@openzeppelin/contracts/utils/introspection/ERC165.sol";

Reviewer 3: I don't know if there's a "canonical" answer, but for things being published in the EIPs repo, my feeling is not to use any external libraries, though for ERCs I wouldn't mind seeing a link to an external repo (I've done that myself), though there are different opinions about this.

> import "@openzeppelin/contracts/utils/introspection/ERC165.sol";

Author: All right, so I've decided to link my personal implementation rather than putting it in the EIP. That way if other implementations come up I can always put them here, and the EIP will be shorter.

    function isDelegateOf(address owner) public view returns (bool) {
        return isDelegate(msg.sender, owner);
    }

    function supportsInterface(bytes4 interfaceId) public view virtual override(NTT) returns (bool) {
        return 
            interfaceId == type(INTTDelegate).interfaceId ||
            super.supportsInterface(interfaceId);
    }

    function _delegateAsDelegateOrCreator(address operator, address owner, bool isCreator) private {
        require(
            isCreator || _allowed[msg.sender][owner],
            "Only contract creator or allowed operator can delegate"
        );
        if (!isCreator) {
            _allowed[msg.sender][owner] = false;
        }
        _allowed[operator][owner] = true;
    }

    function _mintAsDelegateOrCreator(address owner, bool isCreator) private {
        require(
            isCreator || _allowed[msg.sender][owner],
            "Only contract creator or allowed operator can mint"
        );
        if (!isCreator) {
            _allowed[msg.sender][owner] = false;
        }
        _mint(owner);
    }
}
```
<!-- AUTO-GENERATED-CONTENT:END -->

> ## NTT for EIP ?

Reviewer 3: While the idea is novel, this also isn't an EIP-1 section, and should be removed. You could release it as a blog post or similar.
